### PR TITLE
MatZq `dot_product`

### PR DESCRIPTION
**Pull request by jnsiemer**
**Description**
This PR implements...
- [x] dot_product

for `MatZq`.
It is way more convenient than `(vec_a.transpose() * vec_b).get_entry(0,0).unwrap()`.

**Testing**
- [x] I added basic working examples (possibly in doc-comment)
- [x] I added tests for large (pointer representation) values
- [x] I triggered all possible errors in my test in every possible way
- [x] I included tests for all reasonable edge cases

**Checklist:**
- [x] I have performed a self-review of my own code
  - [x] The code provides good readability and maintainability s.t. it fulfills best practices like talking code, modularity, ...
    - [x] The chosen implementation is not more complex than it has to be
  - [x] My code should work as intended and no side effects occur (e.g. memory leaks)
  - [x] The doc comments fit our style guide

### CODE REVIEW

File: src/integer_mod_q/mat_zq/vector/dot_product.rs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+};
+use flint_sys::fmpz::fmpz_addmul;
+
+impl MatZq {
+    /// Returns the dot product of two vectors of type [`MatZq`].
+    ///
+    /// Paramters:
+    /// - `other`: specifies the other vector the dot product is calculated over
+    ///
+    /// Returns the resulting `dot_product` as a [`Zq`] or an error,
```
Comment (MoogSven) [inline src/integer_mod_q/mat_zq/vector/dot_product.rs:25]:
```suggestion
    /// Returns the resulting dot product as a [`Zq`] or an error,
```

File: src/integer_mod_q/mat_zq/vector/dot_product.rs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    error::MathError,
+    integer::Z,
+    integer_mod_q::{MatZq, Zq},
+    traits::{GetNumColumns, GetNumRows},
+};
+use flint_sys::fmpz::fmpz_addmul;
+
+impl MatZq {
+    /// Returns the dot product of two vectors of type [`MatZq`].
+    ///
```
Comment (MoogSven) [inline src/integer_mod_q/mat_zq/vector/dot_product.rs:21]: The two vectors 
```suggestion
    ///
    /// The orientation of the input vectors (row vs. column) is ignored.
    /// As a result, it can be chosen arbitrarily.
    ///
```

Or something similar that say that the different vector types can be combined with each other.

File: src/integer_mod_q/mat_zq/vector/dot_product.rs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                self_entries.len(),
+                other_entries.len()
+            )));
+        }
+
+        // calculate dot product of vectors
+        let mut result = Z::ZERO;
+        for i in 0..self_entries.len() {
+            // sets result = result + self.entry[i] * other.entry[i] without cloned Z element
+            unsafe { fmpz_addmul(&mut result.value, &self_entries[i], &other_entries[i]) }
```
Comment (MoogSven) [inline src/integer_mod_q/mat_zq/vector/dot_product.rs:87]: Nice :+1:

File: src/integer_mod_q/mat_zq/vector/dot_product.rs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    fn large_numbers() {
+        let vec_1 = MatZq::from_str(&format!("[[1,-1,{}]] mod {}", i64::MAX, u64::MAX)).unwrap();
+        let vec_2 = MatZq::from_str(&format!("[[1,{},1]] mod {}", i64::MIN, u64::MAX)).unwrap();
+        let cmp = Z::from(-1) * Z::from(i64::MIN) + Z::from(i64::MAX) + Z::ONE;
+        let cmp = Zq::try_from_z_z(&cmp, &Z::from(u64::MAX)).unwrap();
+
+        let dot_prod = vec_1.dot_product(&vec_2).unwrap();
+
+        assert_eq!(dot_prod, cmp);
+    }
```
Comment (MoogSven) [inline src/integer_mod_q/mat_zq/vector/dot_product.rs:159]: I would suggest to choose the modulus larger so that the result is also large.
Maybe also just one vector with large numbers.
Currently, the result of the dot product is 1 after applying the modulus.

You can also replace `Z::from(-1)` with just `-1` and `Z::ONE` with `1`

**Review by MoogSven — APPROVED** [review]
Looks good, just small changes to comments and to one test case.
